include/i2c.h

```diff
@@ -110,16 +117,66 @@ union __deprecated dev_config {
  * These are for internal use only, so skip these in
  * public documentation.
  */
+struct i2c_slave_config;
+
+typedef int (*i2c_slave_write_requested_cb_t)(
+		struct i2c_slave_config *config);
+typedef int (*i2c_slave_read_requested_cb_t)(
+		struct i2c_slave_config *config, u8_t *val);
+typedef int (*i2c_slave_write_received_cb_t)(
+		struct i2c_slave_config *config, u8_t val);
+typedef int (*i2c_slave_read_processed_cb_t)(
+		struct i2c_slave_config *config, u8_t *val);
+typedef int (*i2c_slave_stop_cb_t)(struct i2c_slave_config *config);
+
+struct i2c_slave_callbacks {
+	/** callback function being called when write is requested */
+	i2c_slave_write_requested_cb_t write_requested;
+	/** callback function being called when read is requested */
+	i2c_slave_read_requested_cb_t read_requested;
+	/** callback function being called when byte has been received */
+	i2c_slave_write_received_cb_t write_received;
+	/** callback function being called when byte has been sent */
+	i2c_slave_read_processed_cb_t read_processed;
+	/** callback function being called when stop occurs on the bus */
+	i2c_slave_stop_cb_t stop;
+};
+
+struct i2c_slave_config {
+	/** Private, do not modify */
+	sys_snode_t node;
+	/** Flags for the slave device defined by I2C_SLAVE_FLAGS_* constants */
+	u8_t flags;
+	/** Address for this slave device */
+	u16_t address;
+	/** Callback functions */
+	const struct i2c_slave_callbacks *callbacks;
+};
+
 typedef int (*i2c_api_configure_t)(struct device *dev,
 				   u32_t dev_config);
 typedef int (*i2c_api_full_io_t)(struct device *dev,
 				 struct i2c_msg *msgs,
 				 u8_t num_msgs,
 				 u16_t addr);
+typedef int (*i2c_api_slave_register_t)(struct device *dev,
+					struct i2c_slave_config *cfg);
+typedef int (*i2c_api_slave_unregister_t)(struct device *dev,
+					  struct i2c_slave_config *cfg);
 
 struct i2c_driver_api {
 	i2c_api_configure_t configure;
 	i2c_api_full_io_t transfer;
+	i2c_api_slave_register_t slave_register;
+	i2c_api_slave_unregister_t slave_unregister;
+};
+
+typedef int (*i2c_slave_api_register_t)(struct device *dev);
+typedef int (*i2c_slave_api_unregister_t)(struct device *dev);
```

Comment:
I can't remember if I asked this already but: is there some use-case for runtime registration/unregistration of i2c slaves? If we compare to actual i2c chip slaves electronically connected: there is no way to add/remove them at runtime, so maybe it's useless to be able to do that for soft-based i2c slaves, don't you think? Or am I missing a different usage for these?

Reply:
Yes, there is one use case when you use an STM32 as Embedded Controller for a board, you can provide different i2c interfaces depending on the main SoC state for example, or save power by disabling the i2c slave when in low power mode.

Reply:
You can move slave logic from one i2c to another? is it still on the same bus electronically speaking?

Second use case is not a proper one. Disabling a device for power saving should be done through the PM API, which one will work on the actual i2c device (the one against which you register the slave). At least I don't see why you would "unregister" a slave: this would have no impact on power level (besides the fact that incoming bytes would not be managed).

Reply:
On STM32, we leave the controller enabled until a slave is attached, detaching a slave will stop the controller.

Reply:
You stop the controller, but it's still getting clocked (and thus powered), my point here is that register/unregister are not functions used for power management, that's all.
Well PM API functions are not implemented in i2c on stm32 (as many other drivers btw, this is not the only one. I wonder if anybody is maintaining this PM API ...)

I am interested about what you said about providing different i2c interfaces depending on the main SoC state: could you detail that one: are the 2 interfaces working on the same bus (the same 2 electric lines)?